### ISSUE TITLE
Space Engineers Torch APi Configuration

### ISSUE THREAD

**Issue by m-leonoff** [CubeCoders/AMPTemplates#530]
# Module Request

Note: Please try setting up a configuration yourself before raising an issue to request a configuration:  ~~https://config.getamp.sh/~~
***There is a newer beta version available at https://iceofwraith.github.io/GenericConfigGen/

## Game Information:

Game Title: Space Engineers Torch API
How to obtain: https://torchapi.com

This is super necessary for well-established space engineer servers. Torch API enables the use of server management plugins.

## I confirm:

- [ x] that I have searched for an existing module request for this application.
- [ x] that the server is an official server distributed by the games publishers/developers.
- [ x] that the server is available to the general public.
- [ x] that the server can be run headless (without a GUI).
- [ x] I have already attempted to create a configuration myself using the configuration tool
- [x] that my checkboxes above look like the one on the left (remove this line when done)

**Comment by PepperJackExA** [comment]
I would also like this to be added

**Comment by m-leonoff** [comment]
So any chances of this coming to life?

**Comment by Greelan** [comment]
Not possible on Windows given the system user that AMP runs under normally on that platform. 

I played around with it on Linux (obv with Wine) and iirc it worked but it's not great since Torch starts a separate process for the server but AMP can't manage it.

**Comment by Bitenode** [comment]
have you tried the nogui mode? i can msg Bishbash777 see if he could help make a workaround it only really needs to have a plugin in your torch server that can communicate with your api

**Comment by southnode** [comment]
So from the Torch github - https://github.com/TorchAPI/Torch/issues/449

This was the original issue with Torch running as the network service just immediately shutting down. Torch could theoretically work with AMP now utilizing whatever development build you already have there @Greelan (if that was the problem in the first place given the NETWORK SERVICE user being non-interactive).

**Comment by Greelan** [comment]
The server complains about being run as a service

**Comment by Greelan** [comment]
Right, I've looked at the commit now. I see they've made a change to address the parent process being a service. Will test if this fixes the previous issue in AMP

**Comment by Greelan** [comment]
Derp, that commit was only 4 days ago and hasn't made it into a build yet

**Comment by Greelan** [comment]
Actually, it seems it has, last build was a day ago.

But still:

`Cannot start service from the command line or a debugger. A Windows Service must first be installed (using installutil.exe) and then started with the ServerExplorer, Windows Services Administrative tool or the NET START command.`

**Comment by sam4413** [comment]
@Greelan We have made several good progress in getting TorchAPI working with AMP.
https://github.com/sam4413/amp-spaceengineers
I have also made a companion torch plugin to go with it to fix the issue of torch spawning a separate process upon startup.
The only downside is that we haven't really managed to get it working with linux / wine, and it only works with Windows Server versions out the box (with windows home/pro having to do extra config), but perhaps we could figure something out?

**Comment by Greelan** [comment]
@sam4413 you prompted me to have another look, since some stuff has changed in AMP over the last year and I now have a Server 2019 VM to test with.

I took some stuff from your repo and adapted it to the current SE template, and made some other improvements and fixes (eg the world/scenario selection now works and a new save can be created). Template in `Greelan/AMPTemplates:dev`

Windows does indeed seem to work well (only tried on Windows Server 2019 so far).

Linux also works, though AMP is not receiving the full console output. Need to continue to investigate that.

I also need to spend some more time looking at the Torch settings, as I am not sure all of them are appropriate for AMP, such as the auto restart settings (since AMP is likely to see that as a crash and try to restart itself, causing a race condition).

**Comment by Greelan** [comment]
Now have console output on Linux, and have gone through and updated the settings to what seem the most appropriate to me.

Would welcome feedback.

**Comment by sam4413** [comment]
That's amazing! I have also done several updates to the torch on amp template, I have removed the LocalPlugins option and Restart on Crash, as those would make amp loose track of its progress, and make it easier to install plugins by dragging and dropping.

Also as mentioned in the discord, we did some large progress for the Torch console, on which you can actually send commands in the console now.

**Comment by evanpurl** [comment]
I got the console part working. We're using a plugin that reads the console from amp, determines if it's a chat message or a command. Sam got the commands functional though.

**Comment by sam4413** [comment]
So for restart on crash, what it does is that at least for torch running windows, it would kill the process and spawn a new one
https://github.com/TorchAPI/Torch/blob/ae51eb9f1937c9be0715108cb5b4f0ea0e4c05d5/Torch.Server/TorchServer.cs#L217

For local plugins, removing the option would just make it so that you have to install all plugins via the Plugins tag by adding <guid>s. Its a nice to have if your developing a custom plugin, but if the plugin is already uploaded onto torch, you can simply grab its guid and load it in that way. It essentially would prevent torch loading extracted folders as plugins, as that is considered a local plugin.

The console part we made a custom plugin called AMP Utilities. It can be found here 
https://github.com/sam4413/AMPUtilities
https://torchapi.com/plugins/view/5c14d8ea-7032-4db1-a2e6-9134ef6cb8d9

For the separate process, it might be because I haven't changed how AMP stops the process, I saw that you have updated it on your end though.


For the discord, we were referring to this message in cubecoders discord https://discord.com/channels/266012086423912458/266015417842139136/1384186454657208402
